### PR TITLE
Feature/capture magewell

### PR DESCRIPTION
**Pull request by PotatoPeeler3000**
This PR allows for viewing videos and logs recorded with a Magewell capture device. Unfortunately videos recorded with Magewell can't be played back with the previous implementation so a specific magewell one had to be made resulting in the following interface `VideoDataPlayer` which has specific implementations for Magewell and Blackmagic.
This follows the logger PR to enable logging with  https://github.com/ihmcrobotics/ihmc-robot-data-logger/pull/8

### CODE REVIEW

File: scs2-session-visualizer-jfx/build.gradle.kts [hunk 1]
```diff
@@ -20,6 +20,10 @@ mainDependencies {
    api("us.ihmc:scs2-session-logger:source")
    api("us.ihmc:scs2-session-visualizer:source")
 
+   api("org.bytedeco:javacv:1.5.10")
```
Comment (PotatoPeeler3000) [inline scs2-session-visualizer-jfx/build.gradle.kts]: Not sure if we need these, need to double check @PotatoPeeler3000

File: scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/BlackMagicVideoDataReader.java
Comment (PotatoPeeler3000) [inline scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/BlackMagicVideoDataReader.java:1]: Renamed because this loads logs that were captured with Blackmagic

File: scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/MagewellDemuxer.java
Comment (PotatoPeeler3000) [inline scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/MagewellDemuxer.java:1]: This is the demuxer that would replace the current one, its updated and cleaner

File: scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/MagewellVideoDataReader.java
Comment (PotatoPeeler3000) [inline scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/MagewellVideoDataReader.java:1]: This is the loader for logs that were captured with Magewell, unfortunately the two capture methods can't be loaded with the same demuxer (not entirely sure why but it didn't work)

File: scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/MultiVideoDataReader.java [hunk 2]
```diff
@@ -27,8 +27,23 @@ public MultiVideoDataReader(File dataDirectory, LogProperties logProperties, Bac
          Camera camera = cameras.get(i);
          try
          {
-            VideoDataReader reader = new VideoDataReader(camera, dataDirectory, logProperties.getVideo().getHasTimebase());
-            readers.add(reader);
+            VideoDataReader reader;
+            if (camera.getType().toString().equals("MAGEWELL_CAPTURE"))
```
Comment (PotatoPeeler3000) [inline scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/MultiVideoDataReader.java]: Depending on the type of capture algorithm used, we can create a differ VideoDataReader to load it correctly

File: scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/MultiVideoDataReader.java [hunk 3]
```diff
@@ -62,7 +77,9 @@ public void crop(File selectedDirectory, long startTimestamp, long endTimestamp,
             progressConsumer.info("Cropping video (%s)".formatted(camera.getVideoFileAsString()));
             double progressPercentage = (double) i / (double) readers.size();
             progressConsumer.progress(progressPercentage);
-            subProgressConsumer = progressConsumer.subProgress("Cropping video (%s): ".formatted(camera.getVideoFileAsString()), progressPercentage, (i + 1.0) / readers.size());
+            subProgressConsumer = progressConsumer.subProgress("Cropping video (%s): ".formatted(camera.getVideoFileAsString()),
```
Comment (PotatoPeeler3000) [inline scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/MultiVideoDataReader.java:79]: Formatting? @PotatoPeeler3000

File: scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/VideoDataReader.java [hunk 4]
```diff
@@ -1,351 +1,39 @@
...
-import java.io.BufferedReader;
 import java.io.File;
-import java.io.FileNotFoundException;
-import java.io.FileReader;
 import java.io.IOException;
-import java.io.PrintWriter;
-import java.util.Arrays;
 
-public class VideoDataReader
+public interface VideoDataReader
```
Comment (PotatoPeeler3000) [inline scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/VideoDataReader.java:12]: This is now an interface for the different VideoDataReaders that are implemented

File: scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/VideoViewer.java [hunk 5]
```diff
@@ -228,7 +231,67 @@ protected void layoutChildren()
 
    public void update()
    {
-      FrameData currentFrameData = reader.pollCurrentFrame();
+      if (reader.getClass().equals(MagewellVideoDataReader.class))
```
Comment (PotatoPeeler3000) [inline scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/VideoViewer.java]: This is ugly, but allows support for the old and new logs, maybe there is a better way to do this @PotatoPeeler3000 @SylvainBertrand

File: scs2-session-visualizer-jfx/src/test/java/us/ihmc/scs2/sessionVisualizer/jfx/session/TimestampScrubberTest.java
Comment (PotatoPeeler3000) [inline scs2-session-visualizer-jfx/src/test/java/us/ihmc/scs2/sessionVisualizer/jfx/session/TimestampScrubberTest.java:1]: Testsssssssssssss, need to add more and check this before merge me thinks @PotatoPeeler3000

**Comment by PotatoPeeler3000** [comment]
This build is failing because its pulling the logger code from the wrong branch, not sure how to fix that rn but it claims it can't find the camera.getType() method which does exist on my branch

File: scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/BlackMagicVideoDataReader.java
Comment (PotatoPeeler3000) [inline scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/BlackMagicVideoDataReader.java:1]: This class wasn't changed, it was renamed so it could be the BlackMagic specific implementation, and it now implements and interface which gets used at a higher level

File: scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/BlackMagicVideoDataReader.java [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+import us.ihmc.robotDataLogger.Camera;
+import us.ihmc.scs2.session.log.ProgressConsumer;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.PrintWriter;
+
+public class BlackMagicVideoDataReader implements VideoDataReader
+{
+   private final TimestampScrubber timestampScrubber;
```
Comment (PotatoPeeler3000) [inline scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/BlackMagicVideoDataReader.java:15]: The TimestampScrubber class was pulled out so it could be used in 2 places and used in tests

File: scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/FrameData.java
Comment (PotatoPeeler3000) [inline scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/FrameData.java:1]: Frame data provides useful information when debugging the timestamps for the videos

File: scs2-session-visualizer-jfx/src/test/java/us/ihmc/scs2/sessionVisualizer/jfx/session/TimestampScrubberTest.java
Comment (PotatoPeeler3000) [reply scs2-session-visualizer-jfx/src/test/java/us/ihmc/scs2/sessionVisualizer/jfx/session/TimestampScrubberTest.java:1]: Maybe do in the future me thinks

File: scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/MultiVideoDataReader.java [hunk 7]
```diff
@@ -27,7 +27,20 @@ public MultiVideoDataReader(File dataDirectory, LogProperties logProperties, Bac
...
+            if (camera.getType().toString().equals("MAGEWELL_CAPTURE"))
+            {
+               reader = new MagewellVideoDataReader(camera, dataDirectory, logProperties.getVideo().getHasTimebase());
+            }
+            else if (camera.getType().toString().equals("CAPTURE_CARD"))
+            {
+               reader = new BlackMagicVideoDataReader(camera, dataDirectory, logProperties.getVideo().getHasTimebase());
+            }
+            else
+            {  // Older logs won't have the camera type set correctly, if there isn't a type set this as the only option
```
Comment (PotatoPeeler3000) [inline scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/MultiVideoDataReader.java:41]: If we don't know the type, we will try to load it with the BlackMagicVideoDataReader but no promises

**Comment by PotatoPeeler3000** [comment]
@PotatoPeeler3000 Need to check cropping a log almost forgot

**Comment by ds58** [comment]
@PotatoPeeler3000 Can you add me back as a reviewer when it's ready?

**Comment by ds58** [comment]
CI is failing, can you check that?

File: scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/BlackMagicVideoDataReader.java [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+
+   public BlackMagicVideoDataReader(Camera camera, File dataDirectory, boolean hasTimeBase) throws IOException
+   {
+      this.camera = camera;
+      name = camera.getNameAsString();
+      boolean interlaced = camera.getInterlaced();
+
+      if (!hasTimeBase)
+      {
+         System.err.println("Video data is using timestamps instead of frame numbers. Falling back to seeking based on timestamp.");
```
Comment (ds58) [inline scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/BlackMagicVideoDataReader.java:33]: I know this was here before but could you change it to use LogTools.error

File: scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/FrameData.java [hunk 9]
```diff
@@ -0,0 +1,16 @@
+package us.ihmc.scs2.sessionVisualizer.jfx.session.log;
+
+import javafx.scene.image.WritableImage;
+
+/**
+ * This class is used for debugging timestamp delays in the videos when viewing them in SCS2. In order to see the timestamp debugging information change the
+ * boolean in {@link VideoViewer#LOGGER_VIDEO_DEBUG} to true. This can also be set as an environmental variable.
+ */
+public class FrameData
```
Comment (ds58) [inline scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/FrameData.java:9]: Can you make this a java record instead of a class?

File: scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/MagewellVideoDataReader.java [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+
+   public MagewellVideoDataReader(Camera camera, File dataDirectory, boolean hasTimeBase) throws IOException
+   {
+      this.camera = camera;
+      name = camera.getNameAsString();
+      boolean interlaced = camera.getInterlaced();
+
+      if (!hasTimeBase)
+      {
+         System.err.println("Video data is using timestamps instead of frame numbers. Falling back to seeking based on timestamp.");
```
Comment (ds58) [inline scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/MagewellVideoDataReader.java:36]: I know this was here before but could you change it to use LogTools.error

**Review by ds58 — CHANGES_REQUESTED** [review]

File: scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/MultiVideoDataReader.java (hunk 2 above)
Comment (rjgriffin42) [reply scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/MultiVideoDataReader.java]: This seems kind of fragile, why would you not compare against the enum CameraType?

File: scs2-simulation/src/main/java/us/ihmc/scs2/simulation/collision/CollisionTools.java [hunk 11]
```diff
@@ -179,7 +179,8 @@ else if (definition instanceof STPRamp3DDefinition stpRamp3DDefinition)
       else if (definition instanceof Ramp3DDefinition ramp3DDefinition)
          return new FrameRamp3D(referenceFrame, toRamp3D(originPose, ramp3DDefinition));
       else if (definition instanceof ModelFileGeometryDefinition modelFileGeometryDefinition)
-         return new FrameConvexPolytope3D(referenceFrame, toConvexPolytope3D(originPose, modelFileGeometryDefinition));
+         return new FrameConvexPolytope3D(referenceFrame, new ConvexPolytope3D());
```
Comment (PotatoPeeler3000) [inline scs2-simulation/src/main/java/us/ihmc/scs2/simulation/collision/CollisionTools.java]: @PotatoPeeler3000 Need to revert this change, shouldn't have been commited

File: scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/MultiVideoDataReader.java (hunk 2 above)
Comment (PotatoPeeler3000) [reply scs2-session-visualizer-jfx/src/main/java/us/ihmc/scs2/sessionVisualizer/jfx/session/log/MultiVideoDataReader.java]: Agreed, fixed that to compare enum types
